### PR TITLE
Add a Stimulus controller to handle scrolling in the back end

### PR DESCRIPTION
**Pull request by zoglo**
_Edited initial PR description (03.01.2024)_

**Feature**
- Replace mootools Backend scroll offset handling with a stimulus controller

____ 

After @aschempp's contribution: https://github.com/contao/contao/pull/6672/commits/5b9e80d24f74434e0982980654984bb3ad8e880f,

you can use following attributes for:

**Saving the Backend Offset** (replaces the previous Backend.getScrollOffset())**
- **focus**:  ```data-action="focus->contao--offset#store"```
- **click**: ```data-action="click->contao--offset#store"```

**Scrolling to a target**
- ```data-contao--offset-target="scrollTo"```

____ 

**Orginal planned feature**

- Adds an auto-scroll to the first selected item in a picker to enhance user experience

**Description**

- Similar approach to https://github.com/contao/contao/issues/6555 but with autoscroll
- Auto-scroll to the first selected item to enhance the user-experience
- No need to scroll and search for the selection anymore

[](https://github.com/contao/contao/assets/55794780/b7288f19-6dad-4652-aa27-1ad6f23ca13f)

**Related**

- [#6662](https://github.com/contao/contao/issues/6662) - Due to layout shift on chrome
- [#6555](https://github.com/contao/contao/issues/6555) - Enhance UX

**Disclaimer**
- Testing this needs you to rebuild the assets again due to the fast pace of commits affecting the backend css and js

### CODE REVIEW

**Comment by zoglo** [comment]
### Check [this](https://github.com/contao/contao/pull/6672#issuecomment-1871639236) comment for new information regarding this PR

___

~~**Update**~~
~~I updated this PR to implement a Stimulus controller that can be used in the future.~~

~~Stimulus [action](https://stimulus.hotwired.dev/reference/actions): Set ScrollOffset~~
```
data-controller="contao--offset" data-action="click->contao--offset#set~~
```

~~Stimulus [target](https://stimulus.hotwired.dev/reference/targets): JumpToTarget~~
```
data-controller="contao--offset data-contao--offset-target="scrollTo"
```

___

~~This PR now also introduces changing the backend offset events~~

~~*Instead of*~~
```
- click events -> 'onclick="Backend.getScrollOffset()"'
- focus events -> 'onfocus="Backend.getScrollOffset()"'
```
~~*you can now use*~~
```
- click events -> 'data-controller="contao--offset" data-action="click->contao--offset#set"'
- focus events -> 'data-controller="contao--offset" data-action="focus->contao--offset#set"'
```
___

~~**Example**~~
~~``<input type="text" data-controller="contao--backend-offset" data-action="focus->contao--backend-offset#set">``~~

~~You are able to change the events to whatever you want using [Stimulus]~~~~(https://stimulus.hotwired.dev/handbook/introduction)~~

~~The latest commit changed the scrollTo to check for the last data-attribute in the DOM by adding these data attributes~~
```
data-contao--offset-target="scrollTo""
```

___

~~**More information**~~
-~~I added comments for when mootools will be removed so we can switch to the new stimulus controller~~

**Comment by zoglo** [comment]
After @aschempp's contribution: https://github.com/contao/contao/pull/6672/commits/5b9e80d24f74434e0982980654984bb3ad8e880f,

you can use following attributes for:

**Saving the Backend Offset** (replaces the previous Backend.getScrollOffset())**
- **focus**:  ```data-action="focus->contao--offset#store"```
- **click**: ```data-action="click->contao--offset#store"```

**Scrolling to a target**
- ```data-contao--offset-target="scrollTo"```

**Comment by aschempp** [comment]
Not sure if we should call this `scroll-offset`?

**Comment by leofeyer** [comment]
Renamed in faa98a5a418044c8d263874ce882a3469828b362.

**Review by leofeyer — CHANGES_REQUESTED** [review]
Unfortunately, I cannot get the feature shown in the video working on my system. The `checked data-contao--scroll-offset-target="scrollTo"` attributes are added to the checkbox, but it is not scrolled into the view automatically.

Also, there are 129 more occurrences of `Backend.getScrollOffset()` in the code base, which causes the deprecation message to be triggered all the time. Shouldn‘t we replace these occurrences in the non-legacy code as well?

**Comment by zoglo** [comment]
I planned to work on this PR again this evening.
I was talking to aschempp on Slack about changing the other occurences but wanted to wait if we should rename the controller as it would affect all the occurences.

Once I've changed the occurences (also the automatic scroll-to), the previously created feature should work as shown in the video @leofeyer

**Comment by leofeyer** [comment]
The "scroll into view" feature works now. 🥳

**Comment by zoglo** [comment]
@leofeyer the only missing occurrences of Backend.getScrollOffset() are within JavaScript right now (I listed them down below).

The last commits did change:
- every onclick event on the a-elements to be a data-action with their event shorthand https://github.com/contao/contao/pull/6672/commits/8927a7ea62d324e7ffae9bdc60e8f5b7173ec480
  - See event shorthand for links: https://stimulus.hotwired.dev/reference/actions#event-shorthand 
- every onfocus event to be a data-action listening to the focus event https://github.com/contao/contao/pull/6672/commits/da09a2cec01181145bb5a5c1e02b4fbd4747df75
  - See actions for more information: https://stimulus.hotwired.dev/reference/actions

___

I don't know how to handle the last occurences without breaking anything, maybe @aschempp could help out :) ?

If it is not possible to call the method within the stimulus controller because a [cross-controller action](https://stimulus.hotwired.dev/reference/controllers#cross-controller-coordination-with-events) needs to be set up, 
I would suggest just adding a simple helper method in normal JavaScript that stores the current offset in the session storage.
___

**Last missing Backend.getScrollOffset() occurences**

[sectionwizard.js](https://github.com/contao/contao/blob/5.x/core-bundle/assets/scripts/sectionwizard.js)
- L46
- L61

[core.js](https://github.com/contao/contao/blob/5.x/core-bundle/assets/scripts/core.js)
- L683
- L749
- L767
- L1021
- L1076
- L1086
- L1166
- L1182
- L1191
- L1210
- L1229
- L1248
- L1394
- L1413
- L1493
- L1509
- L1929

[modulewizard.js](https://github.com/contao/contao/blob/5.x/core-bundle/assets/scripts/modulewizard.js)
- L46
- L78
- L95

[be_ace.html5](https://github.com/contao/contao/blob/5.x/core-bundle/contao/templates/backend/be_ace.html5)
- L75

[be_tinyMCE.html5](https://github.com/contao/contao/blob/5.x/core-bundle/contao/templates/backend/be_tinyMCE.html5)
- L31

**Comment by e-spin** [comment]
> Editet initial PR description (**01.03.2024**)

03.01.2024 or 01/03/2024

**Comment by leofeyer** [comment]
@zoglo I have spoken to @aschempp and he suggests to use a [global event](https://stimulus.hotwired.dev/reference/actions#global-events). Something like this:

```
<body data-controller="contao--scroll-offset" data-action="scroll-offset@window->contao--scroll-offset#save">
```

Then we could trigger the event via `window.fireEvent('scroll-offset')` in the remaining places where we still use `Backend.getScrollOffset()`. Does this make sense to you?

(Maybe the event should be called `save-scroll-offset` instead of just `scroll-offset`.)

**Comment by zoglo** [comment]
@leofeyer, @aschempp 

I did get rid of the Backend.getScrollOffset() completely thanks to your hints.

**However:**

```js 
window.fireEvent('save-scroll-offset')
```
would only fire once and then never trigger again so I am using:
```js
window.dispatchEvent(new CustomEvent('save-scroll-offset'))
```
instead.

I also did bind the global action to `<html>` as the data-controller has already been registered there :)

**Comment by leofeyer** [comment]
Perfect. 👌 Thank you very much!

**Comment by aschempp** [comment]
looks good! Two thoughts only:
 1. should it be `store-scroll-offset` according to the method name? 
 2. we could probably use `new Event(...)` instead of `new CustomEvent(...)` since we don't have any custom data. No idea if that makes a real difference though. https://developer.mozilla.org/en-US/docs/Web/Events/Creating_and_triggering_events#creating_custom_events

Also be aware that this is pure legacy handling! Instead of the `copy` action handler triggering the event, we should actually trigger that action on the element when we refactor the copy action. But that's for a later stage 🙃

**Comment by zoglo** [comment]
I implemented @aschempp's changes and also made sure that the stimulus action for the "scrollOffset" is triggered,

However, there is a **race condition** with the `autofocus.js` (See: https://github.com/contao/contao/pull/4345) that is loaded after the controller action (I also encountered this the first time before I switched to using *stimulus*)

**There are a few approaches to counter this**

- rewrite the `autofocus.js` as a stimulus controller -> It will load first due to the ascending letter (with attributes / without attributes aka *dirty*) 
- implement a switch within `sessionStorage` within `scroll-offset-controller.js` and `autofocus.js`
- load the `scroll-offset-controller.js` (with the connect and afterLoad) after `autofocus.js`

That is something that I don't want do decide on my own :/

**Comment by leofeyer** [comment]
> * rewrite the `autofocus.js` as a stimulus controller -> It will load first due to the ascending letter (with attributes / without attributes aka _dirty_)

What exactly do you mean by "with attributes / without attributes"?

> * load the `scroll-offset-controller.js` (with the connect and afterLoad) after `autofocus.js`

Isn‘t that the case already? See https://github.com/contao/contao/blob/5.x/core-bundle/assets/backend.js

**Comment by fritzmg** [comment]
> > * load the `scroll-offset-controller.js` (with the connect and afterLoad) after `autofocus.js`
> 
> Isn‘t that the case already? See https://github.com/contao/contao/blob/5.x/core-bundle/assets/backend.js

It's not the loading order, but rather the execution order. `autofocus.js` uses `window.addEventListener('DOMContentLoaded')` and thus has a high probability of being executed after the stimulus controllers as these get executed immediately when their elements are added to the DOM.

File: core-bundle/assets/controllers/scroll-offset-controller.js [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+        if (this.offset) {
+            window.scrollTo({
+                top: this.offset + this.additionalOffset,
+                behavior: this.behaviorValue
+            });
+
+            this.offset = null;
+        }
+
+        this.buttons = document.querySelectorAll('.tl_submit_container button[name]:not([name="save"])');
```
Comment (fritzmg) [inline core-bundle/assets/controllers/scroll-offset-controller.js]: Can't we output `data-action` attributes on these buttons instead?

**Review by fritzmg — COMMENTED** [review]
Personally I would also prefer `block: 'center'` for the `.scrollIntoView` calls 🙂

**Comment by leofeyer** [comment]
As discussed in the Contao call, we want to add the autofocus functionality to the scroll offset controller.

File: core-bundle/assets/controllers/scroll-offset-controller.js (hunk 1 above)
Comment (zoglo) [reply core-bundle/assets/controllers/scroll-offset-controller.js]: Most of the buttons are added dynamically through a `buttons_callback`:

- `$arrButtons` within `DC_File`
- `$strButtons` within `DC_Folder`
- `$strButtons` within `DC_Table`

You'd have to check every item within the callback and add/replace it in this case

**Comment by zoglo** [comment]
The last change I made was to make sure that the autofocus works when *no saved scroll offset* is present.

- I did implement https://github.com/contao/contao/pull/6704 into the `scroll-offset-controller`
- The [automatic page location restoration](https://developer.mozilla.org/en-US/docs/Web/API/History/scrollRestoration) has been disabled in the last commit to make sure that the controller handles the (last) position 
- The scrollIntoView option [{block: center}](https://github.com/contao/contao/pull/6672#pullrequestreview-1815714425) was also introduced in the latest commits @fritzmg 

If you are okay with this approach, we could close https://github.com/contao/contao/pull/6704 as well @leofeyer 

--- 

**ToDo:**

- The last missing feature is saving the current target when saving the scroll offset + focusing it on reload (that @aschempp requested on 11.01.2024 in the contao-call)

**Comment by leofeyer** [comment]
> * The last missing feature is saving the current target when saving the scroll offset + focusing it on reload (that @aschempp requested on 11.01.2024 in the contao-call)

This could also be a follow-up PR.

> * The [automatic page location restoration](https://developer.mozilla.org/en-US/docs/Web/API/History/scrollRestoration) has been disabled in the last commit to make sure that the controller handles the (last) position

Is this a good idea? Users would always find themselves at the top of the page when they navigate using the "back" button of their browser.

**Comment by zoglo** [comment]
You are right, we should use this when scroll offset or auto-focus needs to happen as it's mandatory for it to function properly.

**Comment by ausi** [comment]
> You are right, we should use this when scroll offset or auto-focus needs to happen as it's mandatory for it to function properly.

Why is `history.scrollRestoration = 'manual'` mandatory for setting the scroll position?
We want to set the scroll position when navigating forwards, not on using the back button of the browser I think.

**Comment by zoglo** [comment]
> Why is `history.scrollRestoration = 'manual'` mandatory for setting the scroll position? We want to set the scroll position when navigating forwards, not on using the back button of the browser I think.

After implementing the autofocus into the controller, the scroll offset did not work yet due to native scroll restoration.

The question is how to handle this case when refreshing the page.
~~The whole idea of the scroll offset controller is to handle the position, we kinda want to override the browser behavior in this case.~~

*Edit*
- I guess we could remove it completely and just use the dedicated scroll-offset after saving the page

**Review by ausi — CHANGES_REQUESTED** [review]
Autofocus does not work correctly.
In the old implementation the headline field was the one getting focused, now it gets skipped and a field further down gets focused.

<kbd>Save and close</kbd> does not work correctly. It now restores the scroll position of the edit mask on the parent view, which it should not.

**Comment by leofeyer** [comment]
> Autofocus does not work correctly. In the old implementation the headline field was the one getting focused, now it gets skipped and a field further down gets focused.

Fixed in 46533f2a358fd0c777f337acaf9df6da64c523be.

File: core-bundle/assets/controllers/scroll-offset-controller.js [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+                block: this.blockValue
+            });
+
+            this.offset = null;
+        }
+
+        this.buttons = document.querySelectorAll('.tl_submit_container button[name]:not([name="save"])');
+
+        this.buttons.forEach((button) => {
+            button.addEventListener('click', this.store, { passive: true });
```
Comment (leofeyer) [inline core-bundle/assets/controllers/scroll-offset-controller.js]:
```suggestion
            button.addEventListener('click', this.discard, { passive: true });
```

Should this not be `this.discard`, so the offset is removed? (Also in line 86.)

**Review by leofeyer — APPROVED** [review]
Everything works now. 🥳 

The "saving the current target when saving the scroll offset + focusing it on reload" feature is not yet implemented, but I‘d vote to add this in a follow-up PR.

**Review by ausi — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you so much @zoglo for your time and patience! ❤️

**Pull request merged**